cfme/common/provider.py

```diff
@@ -1090,6 +1090,102 @@ def cleanup_vm(vm_name, provider):
         logger.warning('Failed to clean up VM %s on provider %s', vm_name, provider.key)
 
 
+def _is_vm_suspended(provider, vm_name):
+    return provider.mgmt.can_suspend and provider.mgmt.is_vm_suspended(vm_name)
+
+
+def _is_vm_paused(provider, vm_name):
+    return provider.mgmt.can_pause and provider.mgmt.is_vm_paused(vm_name)    
+
+
+def _is_vm_suspended_or_paused(provider, vm_name):
+    return (provider.mgmt.can_suspend and provider.mgmt.is_vm_suspended(vm_name) or
+            provider.mgmt.can_pause and provider.mgmt.is_vm_paused(vm_name))
+
+
+def ensure_vm_state(provider, vm_name, state):
+    """
+    Ensures that the VM/instance is in proper state on provider using provider API.
```

Comment:
This doesn't seem appropriate for the provider module since its all to do with VM modification and state.

There's a method in `common.VM`, wait_for_vm_state_change, and a method in `cfme.cloud.instance.Instance`, wait_for_instance_state_change` that are written to use the UI and wait for the VM to get to given state.

I'm concerned about the diffusion of all these different methods to wait/ensure/control vm/instance power state, with little clarity or construct between how/where/when to use them.

Ideally VM would be using Sentaku, and the existing implementations could be collapsed into a single implementation for UI state changes, and this would be used as the rest implementation of that Sentaku ContextualMethod. But we're not there yet (sentaku for BaseVM) so its not clear the best way to proceed without Sentaku.

@psav thoughts on the placement here in provider module, and the overlap with existing UI methods? I'm thinking moving this method to BaseVM, that only takes a `state` parameter since provider is a VM attribute.

The existing uses could use VM.factory with the provider and vm_name they're passing here to create a VM object where they can call `ensure_vm_state(state)`

Reply:
Yes, I wasn't entirely sure where it would best live either -- I put it here because 
1) the "clean up VM on provider" (`cleanup_vm` right above this set of code) method already lived in here
2) this method only talks to the provider to ensure VM state -- because it is used to ensure a state on the provider side BEFORE testing with the UI. We're not using the UI for this particular method.

The name could be changed to "ensure_vm_state_on_provider" or something like that if we need to clarify its purpose further? But, the fact we import it from cfme.common.provider might also clarify that :)

Reply:
@mshriver did this ^^^ clarify anything for you? :) If we don't have the method in cfme.common.provider, is there a better place? Maybe we need a whole new file in utils or something. We could do a "code shuffle" in another PR?

Reply:
@bsquizz sorry for delay. Both valid points, the use of REST here when there are near identical UI methods in the infra|cloud BaseVM derived classes is what makes sentaku the best option here. This would be a larger refactor than is necessary at this time though, so I think your overall approach is fine.

I would suggest a better place to implement the function would be on BaseVM, with an `ensure_state_on_provider` suffix. The name and provider are available attributes, and won't need to be passed. This provides for calling via `some_test_vm.ensure_state_on_provider(vm.RUNNING)`, where `some_test_vm` can be easily generated with the correct vm/instance type via `BaseVM.factory` with a vm_name and provider object.

Reply:
I can move it there. If we go the route of using the instance attrs I'll need to adjust the test cases using the `verify_vm_*` fixtures because those are all dependant on the vm_name fixture.